packages/lib/slots.ts

```diff
@@ -192,18 +195,24 @@ function buildSlotsWithDateRanges({
       ? range.start
       : startTimeWithMinNotice;
 
-    slotStartTime =
-      slotStartTime.minute() % interval !== 0
-        ? slotStartTime.startOf("hour").add(Math.ceil(slotStartTime.minute() / interval) * interval, "minute")
-        : slotStartTime;
-
     // Adding 1 minute to date ranges that end at midnight to ensure that the last slot is included
     const rangeEnd = range.end
       .add(dayjs().tz(organizerTimeZone).utcOffset(), "minutes")
       .isSame(range.end.endOf("day").add(dayjs().tz(organizerTimeZone).utcOffset(), "minutes"), "minute")
       ? range.end.add(1, "minute")
       : range.end;
 
+    slotStartTime =
+      slotStartTime.minute() % interval !== 0
+        ? showOptimizedSlots
+          ? rangeEnd.diff(slotStartTime, "minutes") % interval > interval - slotStartTime.minute()
+            ? slotStartTime.add(interval - slotStartTime.minute(), "minute")
+            : slotStartTime.add(rangeEnd.diff(slotStartTime, "minutes") % interval, "minute")
```

Comment:
`rangeEnd.diff(slotStartTime, "minutes") % interval` - gives the remaining grace minutes available for adjustment available after consuming maximum possible slots.

Condition `rangeEnd.diff(slotStartTime, "minutes") % interval > interval - slotStartTime.minute()` -
After adding this grace (or free) minutes available for adjustment to `slotStartTime`, if result exceeds the 'next start of the hour` , it is adjusted to 'start of the hour`